### PR TITLE
ENH: Visualize feature table with a heatmap

### LINKED ISSUE #36: feature-table heatmaps would be useful for exploring data

### PR DESCRIPTION
**Pull request by thermokarst**
This is a simple first-cut at heatmap visualization, sans interactivity.

Fixes #36

### CODE REVIEW

**Comment by thermokarst** [comment]
Looks like seaborn isn't showing one tick-mark per sample once you have enough samples in the feature table, will need to investigate next week.

File: q2_feature_table/_heatmap/_visualizer.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        if not table_ids.issubset(metadata_ids):
+            raise ValueError('Missing samples in metadata: %r' %
+                             table_ids.difference(metadata_ids))
+
+    cbar_label = 'frequency'
+    if normalize:
+        table = table.apply(_normalizer)
+        cbar_label = 'log10 frequency'
+
+    if category is not None:
```
Comment (thermokarst) [inline q2_feature_table/_heatmap/_visualizer.py]: I should probably break this whole block out into a testable util function, it is doing a lot of untested jazz. 🎷

**Comment by jairideout** [comment]
I reviewed and tested locally, looks good! I made updates to the code based on my review and put them in a squash commit. See the commit message for details about what changed!

File: q2_feature_table/_heatmap/_visualizer.py [hunk 2]
```diff
@@ -109,15 +110,13 @@ def heatmap(output_dir, table: pd.DataFrame,
     if table.empty:
         raise ValueError('Cannot visualize an empty table.')
 
-    category = metadata.to_series() if metadata is not None else None
-
     # Validation
-    if category is not None:
-        table = _munge_metadata(category, table, cluster)
+    if metadata is not None:
+        table = _munge_metadata(metadata.to_series(), table, cluster)
```
Comment (thermokarst) [inline q2_feature_table/_heatmap/_visualizer.py]: :+1: (separate lines leftover from previous iteration of code)

File: q2_feature_table/_heatmap/_visualizer.py [hunk 3]
```diff
@@ -109,15 +110,13 @@ def heatmap(output_dir, table: pd.DataFrame,
...
 
     cbar_label = 'frequency'
     if normalize:
-        table = table.apply(lambda x: np.log(x + 1))
+        table = table.apply(lambda x: np.log10(x + 1))
```
Comment (thermokarst) [inline q2_feature_table/_heatmap/_visualizer.py]: oops! thanks!

File: q2_feature_table/_heatmap/_visualizer.py [hunk 4]
```diff
@@ -148,6 +147,10 @@ def heatmap(output_dir, table: pd.DataFrame,
     heatmap_plot.ax_col_dendrogram.set_position([col.x0, hm.y0 + height, width,
                                                  col.height])
 
+    # https://stackoverflow.com/a/34697479/3776794
+    plt.setp(heatmap_plot.ax_heatmap.xaxis.get_majorticklabels(), rotation=90)
+    plt.setp(heatmap_plot.ax_heatmap.yaxis.get_majorticklabels(), rotation=0)
```
Comment (thermokarst) [inline q2_feature_table/_heatmap/_visualizer.py]: Fun! Looks like this is related to seaborn matplotlib magick --- if the labels are long enough they are rotated correctly on the y-axis:

screen shot 2017-09-12 at 9 22 35 pm

👍 for forcing it to always be rotated

File: q2_feature_table/_heatmap/assets/index.html [hunk 5]
```diff
@@ -8,7 +8,9 @@
 
 <div class='row'>
   <div class='col-md-12'>
-    <a href='./feature-table-heatmap.svg'>Download SVG</a>
+    <a href='./feature-table-heatmap.svg' target="_blank" rel="noopener noreferrer" class='btn btn-default'>
```
Comment (thermokarst) [inline q2_feature_table/_heatmap/assets/index.html]: 👍

File: q2_feature_table/plugin_setup.py [hunk 6]
```diff
@@ -374,26 +374,29 @@
             q2_feature_table.heatmap_choices['color_scheme']),
     },
     name='Generate a heatmap representation of a feature table',
-    description=('Generate a heatmap representation of a feature table with '
-                 'optional clustering on both the sample and feature axes.'),
+    description='Generate a heatmap representation of a feature table with '
+                'optional clustering on both the sample and feature axes.\n\n'
+                'Tip: To generate a heatmap containing taxonomic annotations, '
```
Comment (thermokarst) [inline q2_feature_table/plugin_setup.py]: MVP - TIP OF THE YEAR

File: q2_feature_table/tests/test_heatmap.py [hunk 7]
```diff
@@ -121,6 +116,17 @@ def test_munge_metadata_simple(self):
                            index=exp_idx)
         assert_frame_equal(exp, obs)
 
+    def test_munge_metadata_ids_different_order(self):
```
Comment (thermokarst) [inline q2_feature_table/tests/test_heatmap.py]: Nice.

File: q2_feature_table/_heatmap/_visualizer.py (hunk 4 above)
Comment (jairideout) [reply q2_feature_table/_heatmap/_visualizer.py]: Interesting! I also noticed there's different behavior across seaborn versions too.

**Pull request merged**